primitives/src/rescue/sponge.rs

```diff
@@ -0,0 +1,441 @@
+                "Rescue sponge Error : input to sponge hashing function is not multiple of RATE."
+                    .to_string(),
+            ));
+        }
+        // ABSORB PHASE
+        let mut r = Self::from_state(RescueVector::zero(), &Permutation::default());
+        r.absorb(&input);
+
+        // SQUEEZE PHASE
+        Ok(r.squeeze_native_field_elements(num_output))
+    }
+}
+
+impl<F: RescueParameter, const CHUNK_SIZE: usize> RescueSponge<F, CHUNK_SIZE> {
+    /// Pseudorandom function for Bls12_381 scalar field. It allows unrestricted
+    /// variable length input and number of output elements
+    pub fn full_state_keyed_sponge_with_padding(
+        key: &F,
+        input: &[F],
+        num_outputs: usize,
+    ) -> Vec<F> {
+        assert_eq!(
+            CHUNK_SIZE, STATE_SIZE,
+            "CHUNK_SIZE must be equal to STATE_SIZE. Perhaps you meant to use a RescueCRH instead?"
+        );
+        let mut padded_input = input.to_vec();
+        padded_input.push(F::one());
+        pad_with_zeros(&mut padded_input, STATE_SIZE);
+        Self::full_state_keyed_sponge_no_padding(key, padded_input.as_slice(), num_outputs)
+            .expect("Bug in JF Primitives : bad padding of input for FSKS construction")
+    }
+
+    /// Pseudorandom function for Bls12_381 scalar field. It allows unrestricted
+    /// variable length input and number of output elements. Return error if
+    /// input is not multiple of STATE_SIZE = 4
+    pub fn full_state_keyed_sponge_no_padding(
+        key: &F,
+        input: &[F],
+        num_outputs: usize,
+    ) -> Result<Vec<F>, RescueError> {
+        assert_eq!(
+            CHUNK_SIZE, STATE_SIZE,
+            "CHUNK_SIZE must be equal to STATE_SIZE. Perhaps you meant to use a RescueCRH instead?"
+        );
+        if input.len() % STATE_SIZE != 0 {
+            return Err(RescueError::ParameterError(
+                "Rescue FSKS PRF Error: input to prf function is not multiple of STATE_SIZE."
+                    .to_string(),
+            ));
+        }
+        // ABSORB PHASE
+        let mut state = RescueVector::zero();
+        state.vec[STATE_SIZE - 1] = *key;
+        let mut r = Self::from_state(state, &Permutation::default());
+        r.absorb(&input);
+        Ok(r.squeeze_native_field_elements(num_outputs))
+    }
+}
+
+impl<F: RescueParameter, const CHUNK_SIZE: usize> SpongeExt for RescueSponge<F, CHUNK_SIZE> {
+    type State = RescueVector<F>;
+
+    fn from_state(state: Self::State, _permutation: &Self::Parameters) -> Self {
+        Self { state }
+    }
+
+    fn into_state(self) -> Self::State {
+        self.state
+    }
+}
+
+impl<T: RescueParameter + PrimeField, const CHUNK_SIZE: usize> CryptographicSponge
+    for RescueSponge<T, CHUNK_SIZE>
+{
+    /// Parameters used by the sponge.
+    type Parameters = Permutation<T>;
+
+    /// Initialize a new instance of the sponge.
+    fn new(_permutation: &Self::Parameters) -> Self {
```

Comment:
How do we support a keyed sponge where the initial state is an input key rather than the default value?

Reply:
Do you mean actually passing a `Permutation` parameter? Is [this](https://github.com/EspressoSystems/jellyfish/pull/146/commits/de0ad5320f9cccdbd076ce290620acfd0487bb89) what you mean?

Reply:
No, I mean to pass an initial `state`. Sometimes the initial state might be a secret rather than a default vector.

Reply:
Ah, for that you can use the `from_state` associated function, on the `SpongeExt` trait.